File: php/tests/ScannerTest.php
<?php

declare(strict_types=1);

// Dependency-free test runner (no PHPUnit needed): run `php tests/ScannerTest.php`.

require __DIR__ . '/../src/Scanner.php';
require __DIR__ . '/../src/Cli.php';

use Envdoctor\Cli;
use Envdoctor\Scanner;

$failures = 0;
function check(bool $cond, string $label): void
{
    global $failures;
    if ($cond) {
        echo "  ok  $label\n";
    } else {
        echo "  FAIL $label\n";
        $failures++;
    }
}

// 1) usage detection + comment stripping
$src = <<<'PHP'
<?php
// getenv("COMMENTED")
# getenv("HASH_COMMENTED")
$db = getenv("DB_URL");
$port = $_ENV["PORT"];
$host = $_SERVER["HOST"];
/* getenv("BLOCK_IGNORED") */
PHP;
$used = Scanner::scanSource('config.php', $src);
$names = array_keys($used);
sort($names);
check($names === ['DB_URL', 'HOST', 'PORT'], 'detects getenv/$_ENV/$_SERVER, ignores comments');

// 2) reconcile missing + unused
$dir = sys_get_temp_dir() . '/envd_php_' . uniqid();
mkdir($dir);
file_put_contents("$dir/.env", "DB_URL=x\nUNUSED_KEY=1\n");
file_put_contents("$dir/app.php", "<?php\ngetenv(\"DB_URL\");\ngetenv(\"NEW_FLAG\");\n");
$findings = Scanner::scan($dir);
$errors = array_map(fn($f) => $f->name, array_filter($findings, fn($f) => $f->severity === 'error'));
$warnings = array_map(fn($f) => $f->name, array_filter($findings, fn($f) => $f->severity === 'warning'));
check(in_array('NEW_FLAG', $errors, true), 'NEW_FLAG reported as error');
check(in_array('UNUSED_KEY', $warnings, true), 'UNUSED_KEY reported as warning');
check(!in_array('DB_URL', $errors, true) && !in_array('DB_URL', $warnings, true), 'DB_URL reconciled');

// 3) duplicate keys within a single file
$dir = sys_get_temp_dir() . '/envd_php_' . uniqid();
mkdir($dir);
file_put_contents("$dir/.env", "DUP=a\nSOLO=1\nDUP=b\n");
file_put_contents("$dir/app.php", "<?php\ngetenv(\"DUP\");\ngetenv(\"SOLO\");\n");
$findings = Scanner::scan($dir);
$dup = null;
foreach ($findings as $f) {
    if ($f->rule === 'duplicates') {
        $dup = $f;
        break;
    }
}
check($dup !== null && $dup->name === 'DUP' && $dup->severity === 'error', 'DUP reported as duplicates error');
check($dup !== null && $dup->message === 'defined 2 times in the same file (lines 1, 3)', 'duplicates message correct');
$soloDup = array_filter($findings, fn($f) => $f->rule === 'duplicates' && $f->name === 'SOLO');
check(count($soloDup) === 0, 'single-definition SOLO not a duplicate');
$dupOther = array_filter($findings, fn($f) => $f->name === 'DUP' && in_array($f->rule, ['unused', 'undefined-in-source'], true));
check(count($dupOther) === 0, 'duplicated-but-used DUP not also unused/undefined');

// 4) public-prefix secret leaks
$dir = sys_get_temp_dir() . '/envd_php_' . uniqid();
mkdir($dir);
file_put_contents("$dir/.env", "NEXT_PUBLIC_API_KEY=x\nVITE_SECRET=y\nPUBLIC_URL=z\nAPI_KEY=w\nPUBLIC_KEY=k\n");
$findings = Scanner::scan($dir);
$flagged = array_map(fn($f) => $f->name, array_filter($findings, fn($f) => $f->rule === 'public-prefix'));
check(in_array('NEXT_PUBLIC_API_KEY', $flagged, true), 'NEXT_PUBLIC_API_KEY flagged');
check(in_array('VITE_SECRET', $flagged, true), 'VITE_SECRET flagged');
check(!in_array('PUBLIC_URL', $flagged, true), 'PUBLIC_URL not flagged');
check(!in_array('API_KEY', $flagged, true), 'bare API_KEY not flagged');
check(!in_array('PUBLIC_KEY', $flagged, true), 'PUBLIC_KEY not flagged');
$ppSev = array_filter($findings, fn($f) => $f->rule === 'public-prefix' && $f->severity !== 'error');
check(count($ppSev) === 0, 'public-prefix findings are errors');

// 5) weak-secret detection
$dir = sys_get_temp_dir() . '/envd_php_' . uniqid();
mkdir($dir);
file_put_contents("$dir/.env", "API_KEY=changeme\nDB_PASSWORD=s3cureLongValue!\nAUTH_TOKEN=\nSECRET_KEY=xxxxxx\nDATABASE_URL=postgres://localhost/db\n");
file_put_contents("$dir/app.php", "<?php\ngetenv(\"API_KEY\");\ngetenv(\"DB_PASSWORD\");\ngetenv(\"AUTH_TOKEN\");\ngetenv(\"SECRET_KEY\");\ngetenv(\"DATABASE_URL\");\n");
$findings = Scanner::scan($dir);
$weak = array_values(array_filter($findings, fn($f) => $f->rule === 'weak-secret'));
$weakNames = array_map(fn($f) => $f->name, $weak);
check(in_array('API_KEY', $weakNames, true), 'API_KEY placeholder value is weak');
check(in_array('AUTH_TOKEN', $weakNames, true), 'AUTH_TOKEN empty value is weak');
check(in_array('SECRET_KEY', $weakNames, true), 'SECRET_KEY xxxxxx value is weak');
check(!in_array('DB_PASSWORD', $weakNames, true), 'strong DB_PASSWORD not weak');
check(!in_array('DATABASE_URL', $weakNames, true), 'non-secret-name DATABASE_URL not weak');
$weakOk = array_reduce($weak, fn($c, $f) => $c && $f->severity === 'warning'
    && $f->message === 'secret-looking variable has a weak or placeholder value', true);
check($weakOk, 'weak-secret findings are warnings with correct message');

// 6) typo detection (DATBASE_URL used vs DATABASE_URL defined)
$dir = sys_get_temp_dir() . '/envd_php_' . uniqid();
mkdir($dir);
file_put_contents("$dir/.env", "DATABASE_URL=x\n");
file_put_contents("$dir/app.php", "<?php\ngetenv(\"DATBASE_URL\");\n");
$findings = Scanner::scan($dir);
$typo = null;
foreach ($findings as $f) {
    if ($f->rule === 'typo') {
        $typo = $f;
        break;
    }
}
check($typo !== null && $typo->name === 'DATBASE_URL' && $typo->severity === 'warning', 'DATBASE_URL typo reported');
check($typo !== null && $typo->message === '"DATBASE_URL" is not defined; did you mean "DATABASE_URL"?', 'typo message correct');
$undef = array_filter($findings, fn($f) => $f->rule === 'undefined-in-source' && $f->name === 'DATBASE_URL');
check(count($undef) === 1, 'undefined-in-source still fires for typo name');

// 7) environment-diff detection
$dir = sys_get_temp_dir() . '/envd_php_' . uniqid();
mkdir($dir);
file_put_contents("$dir/.env", "SHARED=1\nONLY_DEFAULT=1\n");
file_put_contents("$dir/.env.production", "SHARED=1\nONLY_PROD=1\n");
$findings = Scanner::scan($dir);
$diffs = array_values(array_filter($findings, fn($f) => $f->rule === 'environment-diff'));
$od = null;
foreach ($diffs as $f) {
    if ($f->name === 'ONLY_DEFAULT') {
        $od = $f;
    }
}
check($od !== null && $od->severity === 'warning' && $od->message === 'defined in default but missing in production', 'ONLY_DEFAULT env-diff correct');
$prodDiff = array_filter($diffs, fn($f) => $f->name === 'ONLY_PROD' && $f->message === 'defined in production but missing in default');
check(count($prodDiff) === 1, 'ONLY_PROD env-diff correct');
$sharedDiff = array_filter($diffs, fn($f) => $f->name === 'SHARED');
check(count($sharedDiff) === 0, 'SHARED present in both → no env-diff');

// 8) type-mismatch detection
$dir = sys_get_temp_dir() . '/envd_php_' . uniqid();
mkdir($dir);
file_put_contents("$dir/.env", "PORT=8080\nWORKERS=4\n");
file_put_contents("$dir/.env.production", "PORT=production\nWORKERS=8\n");
$findings = Scanner::scan($dir);
$tm = array_values(array_filter($findings, fn($f) => $f->rule === 'type-mismatch'));
$tmNames = array_map(fn($f) => $f->name, $tm);
check(in_array('PORT', $tmNames, true), 'PORT integer vs string → type-mismatch error');
check(!in_array('WORKERS', $tmNames, true), 'WORKERS two integers → no type-mismatch');
$portTm = array_filter($tm, fn($f) => $f->name === 'PORT' && $f->severity === 'error'
    && $f->message === 'inferred type differs across environments');
check(count($portTm) === 1, 'PORT type-mismatch is error with correct message');

// 9) JSON shape + no value leak
$dir = sys_get_temp_dir() . '/envd_php_' . uniqid();
mkdir($dir);
file_put_contents("$dir/.env", "API_KEY=changeme\nUNUSED_KEY=supersecretvalue\n");
file_put_contents("$dir/app.php", "<?php\ngetenv(\"API_KEY\");\n");
$findings = Scanner::scan($dir);
$jsonStr = Scanner::toJsonArray($findings);
$arr = json_decode($jsonStr, true);
check(is_array($arr), 'JSON output decodes to an array');
$shapeOk = true;
foreach ($arr as $obj) {
    $keys = array_keys($obj);
    sort($keys);
    if ($keys !== ['file', 'line', 'message', 'name', 'rule', 'severity']) {
        $shapeOk = false;
    }
}
check($shapeOk, 'each JSON object has exactly the required keys');
check(!str_contains($jsonStr, 'supersecretvalue') && !str_contains($jsonStr, 'changeme'), 'no value strings leak into JSON');


// 10) Docker Compose / GitHub Actions / Kubernetes source scanning
$dir = sys_get_temp_dir() . '/envd_php_' . uniqid();
mkdir($dir);
file_put_contents("$dir/.env", "DB_URL=postgres://localhost/db\n");
file_put_contents("$dir/docker-compose.yml", "services:\n  web:\n    environment:\n      SECRET: \${COMPOSE_SECRET}\n      DATABASE: \${DB_URL}\n");
mkdir("$dir/.github", 0777, true);
mkdir("$dir/.github/workflows", 0777, true);
file_put_contents("$dir/.github/workflows/ci.yml", "jobs:\n  deploy:\n    steps:\n      - env:\n          KEY: \${{ secrets.DEPLOY_KEY }}\n          REGION: \${{ vars.REGION }}\n");
file_put_contents("$dir/deployment.yaml", "apiVersion: apps/v1\nkind: Deployment\nspec:\n  value: \${K8S_VAR}\n");
$findings = Scanner::scan($dir);
$undef = array_values(array_filter($findings, fn($f) => $f->rule === 'undefined-in-source'));
$undefNames = array_map(fn($f) => $f->name, $undef);
check(in_array('COMPOSE_SECRET', $undefNames, true), 'COMPOSE_SECRET referenced from compose');
check(in_array('DEPLOY_KEY', $undefNames, true), 'DEPLOY_KEY referenced from actions');
check(in_array('REGION', $undefNames, true), 'REGION referenced from actions vars');
check(in_array('K8S_VAR', $undefNames, true), 'K8S_VAR referenced from k8s manifest');
$msgOk = array_reduce($undef, fn($c, $f) => $c && $f->message === 'referenced but not defined in any environment file', true);
check($msgOk, 'undefined-in-source uses new message');
$dbUnused = array_filter($findings, fn($f) => $f->rule === 'unused' && $f->name === 'DB_URL');
check(count($dbUnused) === 0, 'DB_URL referenced by compose → not unused');
$jsonStr = Scanner::toJsonArray($findings);
check(!str_contains($jsonStr, 'postgres://localhost/db'), 'no infra values leak into JSON');

// diff + sync subcommands
$dir2 = sys_get_temp_dir() . '/envd_php_sub_' . uniqid();
mkdir($dir2);
file_put_contents("$dir2/.env", "A=1\nB=2\n");
file_put_contents("$dir2/.env.production", "A=9\n");
$d = Scanner::diffLabels($dir2, 'default', 'production');
check($d === ['onlyInA' => ['B'], 'onlyInB' => [], 'common' => ['A']], 'diff default vs production');
check(Scanner::syncLabels($dir2, 'default', 'production', true) === ['B'], 'sync --dry-run reports B');
check(!str_contains(file_get_contents("$dir2/.env.production"), 'B='), 'dry-run does not write');
check(Scanner::syncLabels($dir2, 'default', 'production') === ['B'], 'sync appends B');
$prod = file_get_contents("$dir2/.env.production");
check(str_contains($prod, "B=\n") && str_contains($prod, 'A=9') && !str_contains($prod, 'B=2'), 'sync copies no value');

// init / fix generation
$EXAMPLE = "# Generated by envdoctor. Fill in values; do not commit secrets.\nDB_URL=\nPORT=\n";
$gdir = sys_get_temp_dir() . '/envd_php_gen_' . uniqid();
mkdir($gdir);
file_put_contents("$gdir/.env", "DB_URL=secretvalue\n");
file_put_contents("$gdir/app.php", "<?php\ngetenv(\"PORT\");\n");

check(Scanner::envExampleContent($gdir) === $EXAMPLE, '.env.example content exact, no values');
$doc = Scanner::environmentDocContent($gdir);
check(str_contains($doc, '| DB_URL | yes | no |'), 'doc has DB_URL defined/used row');
check(str_contains($doc, '| PORT | no | yes |'), 'doc has PORT defined/used row');
check(!str_contains($doc, 'secretvalue') && !str_contains(Scanner::envExampleContent($gdir), 'secretvalue'), 'no values leak into generated files');

// init: creates when absent
ob_start();
Cli::run(['envdoctor', 'init', '-d', $gdir]);
$out = ob_get_clean();
check(str_contains($out, 'created .env.example') && str_contains($out, 'created ENVIRONMENT.md'), 'init creates both files');
check(file_get_contents("$gdir/.env.example") === $EXAMPLE, 'init wrote exact .env.example');

// init: skips when present, no --force
file_put_contents("$gdir/.env.example", "TOUCHED\n");
ob_start();
Cli::run(['envdoctor', 'init', '-d', $gdir]);
$out = ob_get_clean();
check(str_contains($out, 'skipped .env.example (exists)'), 'init skips existing without --force');
check(file_get_contents("$gdir/.env.example") === "TOUCHED\n", 'init did not overwrite without --force');

// init --force: rewrites
ob_start();
Cli::run(['envdoctor', 'init', '-d', $gdir, '--force']);
$out = ob_get_clean();
check(str_contains($out, 'wrote .env.example'), 'init --force rewrites');
check(file_get_contents("$gdir/.env.example") === $EXAMPLE, 'init --force restored exact content');

// fix: always rewrites
file_put_contents("$gdir/.env.example", "STALE\n");
ob_start();
Cli::run(['envdoctor', 'fix', '-d', $gdir]);
$out = ob_get_clean();
check(str_contains($out, 'wrote .env.example') && str_contains($out, 'wrote ENVIRONMENT.md'), 'fix rewrites both files');
check(file_get_contents("$gdir/.env.example") === $EXAMPLE, 'fix restored exact .env.example');


// schema-validation
$dir3 = sys_get_temp_dir() . '/envd_php_schema_' . uniqid();
mkdir($dir3);
file_put_contents("$dir3/.env", "PORT=99999\nLEVEL=verbose\nAPI=ftp://x\nGOOD=info\n");
file_put_contents("$dir3/envdoctor.schema.json", '{"PORT":{"type":"integer","max":65535},"LEVEL":{"enum":["debug","info"]},"API":{"type":"url"},"MISSING":{"type":"string"},"GOOD":{"enum":["info","warn"]}}');
$sf = [];
foreach (Scanner::scan($dir3) as $f) { if ($f->rule === 'schema-validation') { $sf[$f->name] = $f->message; } }
check(($sf['PORT'] ?? '') === 'value exceeds the maximum', 'schema PORT max');
check(($sf['LEVEL'] ?? '') === 'value is not one of the allowed values', 'schema LEVEL enum');
check(($sf['API'] ?? '') === 'value does not match schema type url', 'schema API type');
check(($sf['MISSING'] ?? '') === 'required by schema but not defined', 'schema MISSING required');
check(!isset($sf['GOOD']), 'schema GOOD passes');

echo $failures === 0 ? "\nAll tests passed\n" : "\n$failures test(s) failed\n";
exit($failures === 0 ? 0 : 1);

File: go/internal/scanner/scanner_test.go
package scanner

import (
	"os"
	"path/filepath"
	"strings"
	"testing"
)

func TestScanSourceDetectsUsage(t *testing.T) {
	src := `package main
import "os"
// os.Getenv("COMMENTED")
func main() {
	_ = os.Getenv("DB_URL")
	_, _ = os.LookupEnv("PORT")
	/* os.Getenv("BLOCK_IGNORED") */
}
`
	used := ScanSource("main.go", src)
	if _, ok := used["DB_URL"]; !ok {
		t.Fatal("expected DB_URL")
	}
	if _, ok := used["PORT"]; !ok {
		t.Fatal("expected PORT")
	}
	if _, ok := used["COMMENTED"]; ok {
		t.Fatal("COMMENTED should be ignored (line comment)")
	}
	if _, ok := used["BLOCK_IGNORED"]; ok {
		t.Fatal("BLOCK_IGNORED should be ignored (block comment)")
	}
}

func TestScanReconciles(t *testing.T) {
	dir := t.TempDir()
	must(t, filepath.Join(dir, ".env"), "DB_URL=x\nUNUSED_KEY=1\n")
	must(t, filepath.Join(dir, "main.go"), "package main\nimport \"os\"\nfunc main(){ os.Getenv(\"DB_URL\"); os.Getenv(\"NEW_FLAG\") }\n")

	res, err := Scan(dir)
	if err != nil {
		t.Fatal(err)
	}
	errs := names(res.Errors())
	warns := names(res.Warnings())
	if !errs["NEW_FLAG"] {
		t.Fatalf("expected NEW_FLAG error, got %v", errs)
	}
	if !warns["UNUSED_KEY"] {
		t.Fatalf("expected UNUSED_KEY warning, got %v", warns)
	}
	if errs["DB_URL"] || warns["DB_URL"] {
		t.Fatal("DB_URL should be reconciled")
	}
}

func TestDuplicates(t *testing.T) {
	dir := t.TempDir()
	must(t, filepath.Join(dir, ".env"), "DB_URL=a\nDB_URL=b\nSOLO=1\n")
	must(t, filepath.Join(dir, "main.go"), "package main\nimport \"os\"\nfunc main(){ os.Getenv(\"DB_URL\"); os.Getenv(\"SOLO\") }\n")

	res, err := Scan(dir)
	if err != nil {
		t.Fatal(err)
	}
	var dups []Finding
	for _, f := range res.Findings {
		if f.Rule == "duplicates" {
			dups = append(dups, f)
		}
	}
	if len(dups) != 1 {
		t.Fatalf("expected 1 duplicate finding, got %d: %v", len(dups), dups)
	}
	if dups[0].Name != "DB_URL" || dups[0].Severity != "error" {
		t.Fatalf("unexpected duplicate finding: %+v", dups[0])
	}
	if want := "lines 1, 2"; !contains(dups[0].Message, want) {
		t.Fatalf("message %q should contain %q", dups[0].Message, want)
	}
	// First occurrence reconciles: DB_URL is neither undefined nor unused.
	if warns := names(res.Warnings()); warns["DB_URL"] {
		t.Fatal("DB_URL should be reconciled, not unused")
	}
}

func TestPublicPrefix(t *testing.T) {
	dir := t.TempDir()
	must(t, filepath.Join(dir, ".env"), "NEXT_PUBLIC_API_KEY=x\nPUBLIC_URL=x\nAPI_KEY=x\n")
	must(t, filepath.Join(dir, "main.go"), "package main\nfunc main(){}\n")

	res, err := Scan(dir)
	if err != nil {
		t.Fatal(err)
	}
	pp := map[string]bool{}
	for _, f := range res.Findings {
		if f.Rule == "public-prefix" {
			if f.Severity != "error" {
				t.Fatalf("public-prefix should be error, got %s", f.Severity)
			}
			pp[f.Name] = true
		}
	}
	if !pp["NEXT_PUBLIC_API_KEY"] {
		t.Fatalf("expected NEXT_PUBLIC_API_KEY flagged, got %v", pp)
	}
	if pp["PUBLIC_URL"] {
		t.Fatal("PUBLIC_URL should not be flagged (not secret-like)")
	}
	if pp["API_KEY"] {
		t.Fatal("API_KEY should not be flagged (no public prefix)")
	}
}

func TestInferType(t *testing.T) {
	cases := map[string]string{
		"":          "empty",
		"3000":      "integer",
		"-7":        "integer",
		"1.5":       "float",
		"TRUE":      "boolean",
		"https://x": "url",
		`{"a":1}`:   "json",
		"hello":     "string",
	}
	for in, want := range cases {
		if got := InferType(in); got != want {
			t.Fatalf("InferType(%q)=%q want %q", in, got, want)
		}
	}
}

func TestLevenshtein(t *testing.T) {
	if Levenshtein("abc", "abc") != 0 {
		t.Fatal("equal strings should be 0")
	}
	if Levenshtein("DATBASE_URL", "DATABASE_URL") != 1 {
		t.Fatal("expected distance 1")
	}
}

func TestWeakSecret(t *testing.T) {
	dir := t.TempDir()
	must(t, filepath.Join(dir, ".env"), "API_KEY=changeme\nSTRONG_TOKEN=s0m3-l0ng-r4nd0m\nSHORT_SECRET=abc\n")
	must(t, filepath.Join(dir, "main.go"), "package main\nfunc main(){}\n")

	res, err := Scan(dir)
	if err != nil {
		t.Fatal(err)
	}
	weak := map[string]bool{}
	for _, f := range res.Findings {
		if f.Rule == "weak-secret" {
			weak[f.Name] = true
		}
	}
	if !weak["API_KEY"] {
		t.Fatal("API_KEY (placeholder) should be weak")
	}
	if !weak["SHORT_SECRET"] {
		t.Fatal("SHORT_SECRET (too short) should be weak")
	}
	if weak["STRONG_TOKEN"] {
		t.Fatal("STRONG_TOKEN should not be weak")
	}
	// No value must leak into any message.
	for _, f := range res.Findings {
		if strings.Contains(f.Message, "changeme") || strings.Contains(f.Message, "s0m3") {
			t.Fatalf("value leaked into message: %q", f.Message)
		}
	}
}

func TestTypoSuggestion(t *testing.T) {
	dir := t.TempDir()
	must(t, filepath.Join(dir, ".env"), "DATABASE_URL=postgres://x\n")
	must(t, filepath.Join(dir, "main.go"), "package main\nimport \"os\"\nfunc main(){ os.Getenv(\"DATBASE_URL\") }\n")

	res, err := Scan(dir)
	if err != nil {
		t.Fatal(err)
	}
	var typos []Finding
	for _, f := range res.Findings {
		if f.Rule == "typo" {
			typos = append(typos, f)
		}
	}
	if len(typos) != 1 || typos[0].Name != "DATBASE_URL" {
		t.Fatalf("expected one typo for DATBASE_URL, got %v", typos)
	}
	if !strings.Contains(typos[0].Message, `did you mean "DATABASE_URL"`) {
		t.Fatalf("unexpected message %q", typos[0].Message)
	}
	// undefined-in-source still fires too.
	if !names(res.Errors())["DATBASE_URL"] {
		t.Fatal("undefined-in-source should still fire for DATBASE_URL")
	}
}

func TestEnvironmentDiff(t *testing.T) {
	dir := t.TempDir()
	must(t, filepath.Join(dir, ".env"), "SHARED=1\n")
	must(t, filepath.Join(dir, ".env.production"), "SHARED=1\nONLY_PROD=1\n")
	must(t, filepath.Join(dir, "main.go"), "package main\nfunc main(){}\n")

	res, err := Scan(dir)
	if err != nil {
		t.Fatal(err)
	}
	diffs := map[string]Finding{}
	for _, f := range res.Findings {
		if f.Rule == "environment-diff" {
			diffs[f.Name] = f
		}
	}
	f, ok := diffs["ONLY_PROD"]
	if !ok {
		t.Fatalf("expected ONLY_PROD diff, got %v", diffs)
	}
	if !strings.Contains(f.Message, "production") || !strings.Contains(f.Message, "default") {
		t.Fatalf("unexpected diff message %q", f.Message)
	}
	if _, ok := diffs["SHARED"]; ok {
		t.Fatal("SHARED present in both, should not diff")
	}
}

func TestTypeMismatch(t *testing.T) {
	dir := t.TempDir()
	must(t, filepath.Join(dir, ".env"), "PORT=3000\nHOST=8080\n")
	must(t, filepath.Join(dir, ".env.production"), "PORT=abc\nHOST=9090\n")
	must(t, filepath.Join(dir, "main.go"), "package main\nfunc main(){}\n")

	res, err := Scan(dir)
	if err != nil {
		t.Fatal(err)
	}
	mm := map[string]Finding{}
	for _, f := range res.Findings {
		if f.Rule == "type-mismatch" {
			mm[f.Name] = f
		}
	}
	if _, ok := mm["PORT"]; !ok {
		t.Fatal("PORT should mismatch (integer vs string)")
	}
	if mm["PORT"].Severity != "error" {
		t.Fatal("type-mismatch should be error")
	}
	if _, ok := mm["HOST"]; ok {
		t.Fatal("HOST integer vs integer should not mismatch")
	}
	if strings.Contains(mm["PORT"].Message, "3000") || strings.Contains(mm["PORT"].Message, "abc") {
		t.Fatal("value leaked into type-mismatch message")
	}
}

func TestInfraSourcesScanned(t *testing.T) {
	dir := t.TempDir()
	must(t, filepath.Join(dir, ".env"), "DB_URL=postgres://x\n")
	must(t, filepath.Join(dir, "docker-compose.yml"),
		"services:\n  web:\n    image: x\n    environment:\n      - TOKEN=${COMPOSE_SECRET}\n      - DB=${DB_URL}\n")
	wfDir := filepath.Join(dir, ".github", "workflows")
	if err := os.MkdirAll(wfDir, 0o755); err != nil {
		t.Fatal(err)
	}
	must(t, filepath.Join(wfDir, "ci.yml"),
		"jobs:\n  build:\n    steps:\n      - run: deploy\n        env:\n          KEY: ${{ secrets.DEPLOY_KEY }}\n          REGION: ${{ vars.REGION }}\n")
	k8sDir := filepath.Join(dir, "k8s")
	if err := os.MkdirAll(k8sDir, 0o755); err != nil {
		t.Fatal(err)
	}
	must(t, filepath.Join(k8sDir, "deploy.yaml"),
		"apiVersion: apps/v1\nkind: Deployment\nspec:\n  value: ${K8S_VAR}\n")

	res, err := Scan(dir)
	if err != nil {
		t.Fatal(err)
	}
	undefined := map[string]bool{}
	for _, f := range res.Findings {
		if f.Rule == "undefined-in-source" {
			undefined[f.Name] = true
			if f.Message != "referenced but not defined in any environment file" {
				t.Fatalf("unexpected message %q", f.Message)
			}
		}
	}
	for _, name := range []string{"COMPOSE_SECRET", "DEPLOY_KEY", "REGION", "K8S_VAR"} {
		if !undefined[name] {
			t.Fatalf("expected %s referenced-but-undefined, got %v", name, undefined)
		}
	}
	// DB_URL is defined and referenced in compose -> not unused.
	for _, f := range res.Findings {
		if f.Rule == "unused" && f.Name == "DB_URL" {
			t.Fatal("DB_URL should be reconciled via compose reference")
		}
	}
	// No values leak.
	for _, f := range res.Findings {
		if strings.Contains(f.Message, "postgres") {
			t.Fatalf("value leaked: %q", f.Message)
		}
	}
}

func TestScanInfraActionsAndEscaping(t *testing.T) {
	// $$ is escaped and must not yield a variable reference.
	used := ScanInfra("compose.yml", "a: $$NOT_VAR\nb: ${REAL_VAR}\n", "compose")
	if _, ok := used["NOT_VAR"]; ok {
		t.Fatal("escaped $$ should not produce a reference")
	}
	if _, ok := used["REAL_VAR"]; !ok {
		t.Fatal("expected REAL_VAR")
	}
	// Actions secret/var/env refs only apply to actions kind.
	compose := ScanInfra("compose.yml", "x: ${{ secrets.ONLY_ACTIONS }}\n", "compose")
	if _, ok := compose["ONLY_ACTIONS"]; ok {
		t.Fatal("secrets. ref should not match for compose kind")
	}
	actions := ScanInfra("ci.yml", "x: ${{ secrets.ONLY_ACTIONS }}\n", "actions")
	if _, ok := actions["ONLY_ACTIONS"]; !ok {
		t.Fatal("secrets. ref should match for actions kind")
	}
}

func contains(s, sub string) bool { return strings.Contains(s, sub) }

func must(t *testing.T, path, content string) {
	t.Helper()
	if err := os.WriteFile(path, []byte(content), 0o644); err != nil {
		t.Fatal(err)
	}
}

func names(fs []Finding) map[string]bool {
	m := map[string]bool{}
	for _, f := range fs {
		m[f.Name] = true
	}
	return m
}

func TestGenerateDocsExactContent(t *testing.T) {
	dir := t.TempDir()
	must(t, filepath.Join(dir, ".env"), "DB_URL=postgres://x\n")
	must(t, filepath.Join(dir, "main.go"), "package main\nimport \"os\"\nfunc main(){ os.Getenv(\"PORT\") }\n")

	docs, err := GenerateDocs(dir)
	if err != nil {
		t.Fatal(err)
	}
	wantExample := "# Generated by envdoctor. Fill in values; do not commit secrets.\nDB_URL=\nPORT=\n"
	if docs[".env.example"] != wantExample {
		t.Fatalf(".env.example = %q, want %q", docs[".env.example"], wantExample)
	}
	md := docs["ENVIRONMENT.md"]
	if !strings.HasPrefix(md, "# Environment variables\n\n| Variable | Defined | Used |\n| --- | --- | --- |\n") {
		t.Fatalf("ENVIRONMENT.md header wrong: %q", md)
	}
	if !strings.Contains(md, "| DB_URL | yes | no |") || !strings.Contains(md, "| PORT | no | yes |") {
		t.Fatalf("ENVIRONMENT.md rows wrong: %q", md)
	}
	if !strings.HasSuffix(md, "\n") {
		t.Fatal("ENVIRONMENT.md should end with newline")
	}
	// Values must never leak.
	if strings.Contains(docs[".env.example"], "postgres") || strings.Contains(md, "postgres") {
		t.Fatal("value leaked into generated docs")
	}
}

func TestDiffAndSync(t *testing.T) {
	dir := t.TempDir()
	must(t, filepath.Join(dir, ".env"), "A=1\nB=2\n")
	must(t, filepath.Join(dir, ".env.production"), "A=9\n")

	d, err := DiffLabels(dir, "default", "production")
	if err != nil {
		t.Fatal(err)
	}
	if len(d.OnlyInA) != 1 || d.OnlyInA[0] != "B" || len(d.OnlyInB) != 0 || len(d.Common) != 1 || d.Common[0] != "A" {
		t.Fatalf("unexpected diff: %+v", d)
	}

	// dry-run does not write
	m, _ := SyncLabels(dir, "default", "production", true)
	if len(m) != 1 || m[0] != "B" {
		t.Fatalf("dry-run missing: %v", m)
	}
	data, _ := os.ReadFile(filepath.Join(dir, ".env.production"))
	if strings.Contains(string(data), "B=") {
		t.Fatal("dry-run should not write")
	}

	// real sync appends B= (no value), keeps A's value
	SyncLabels(dir, "default", "production", false)
	data, _ = os.ReadFile(filepath.Join(dir, ".env.production"))
	if !strings.Contains(string(data), "B=\n") || !strings.Contains(string(data), "A=9") || strings.Contains(string(data), "B=2") {
		t.Fatalf("bad synced file: %q", string(data))
	}
}

func TestSchemaValidation(t *testing.T) {
	dir := t.TempDir()
	must(t, filepath.Join(dir, ".env"), "PORT=99999\nLEVEL=verbose\nAPI=ftp://x\nGOOD=info\n")
	must(t, filepath.Join(dir, "envdoctor.schema.json"),
		`{"PORT":{"type":"integer","max":65535},"LEVEL":{"enum":["debug","info"]},"API":{"type":"url"},"MISSING":{"type":"string"},"GOOD":{"enum":["info","warn"]}}`)
	res, _ := Scan(dir)
	got := map[string]string{}
	for _, f := range res.Findings {
		if f.Rule == "schema-validation" {
			got[f.Name] = f.Message
		}
	}
	want := map[string]string{
		"PORT":    "value exceeds the maximum",
		"LEVEL":   "value is not one of the allowed values",
		"API":     "value does not match schema type url",
		"MISSING": "required by schema but not defined",
	}
	if len(got) != len(want) {
		t.Fatalf("got %v", got)
	}
	for k, v := range want {
		if got[k] != v {
			t.Fatalf("%s: got %q want %q", k, got[k], v)
		}
	}
	if _, ok := got["GOOD"]; ok {
		t.Fatal("GOOD should pass")
	}
}
